add create subscription

fanamoner opened the pull request.

fanamoner committed:
add create subscription

backmeupplz commented on .prettierrc.json:
```diff
@@ -0,0 +1,4 @@
+{
```
Возьми вот этот: https://github.com/Borodutch/frontend-starter-react/blob/main/.prettierrc.json

backmeupplz commented on src/models/Subscription.ts:
```diff
@@ -0,0 +1,47 @@
+import { prop, getModelForClass } from '@typegoose/typegoose'
+const Wallet = require('ethereumjs-wallet').default
```
Можем найти типы получше, чтобы не нужно было делать `require`? `Web3` нам тут разве будет недостаточно?

backmeupplz commented on package.json:
```diff
@@ -23,6 +23,7 @@
   "dependencies": {
     "@typegoose/typegoose": "^7.6.1",
     "dotenv": "^10.0.0",
+    "ethereumjs-wallet": "^1.0.1",
```
Попробуй со `web3` поработать

backmeupplz commented on src/handlers/handleStart.ts:
```diff
@@ -26,5 +27,5 @@ export async function sendStart(ctx: Context) {
   }
 
-  return ctx.replyWithHTML(ctx.i18n.t('help'))
+  return getOrCreateSubscription(ctx)
```
Тут в `startPayload` может прийти ID канала, на который человек пытается подписаться. Плюс, если нет языка, может начаться `sendLanguage`, куда тоже нужно закидывать startPayload — подписка после выбора языка сработает?

backmeupplz commented on src/models/Subscription.ts:
```diff
@@ -0,0 +1,47 @@
+  @prop({ required: true })
+  username: string
+  @prop({ required: true })
+  chatId: string
```
давай сделаем тут `number`

backmeupplz commented on src/models/Subscription.ts:
```diff
@@ -0,0 +1,47 @@
+  @prop({ required: true })
+  chatId: string
+  @prop({ required: true, unique: true })
+  adress: string
```
`address`

backmeupplz replied:
Давай, кстати, укажем именно `ethAddress` тут, или добавим прямо структуру-объект с разными способами оплаты? После ETH мы будем подключать другие криптовалюты.

backmeupplz commented on src/models/Subscription.ts:
```diff
@@ -0,0 +1,47 @@
+  @prop({ required: true, unique: true })
+  privateKey: string
+  @prop({ required: true })
+  payment: number
```
Давай создадим отдельную модель `Payment` и будем его сохранять в бд. У Payment будет поле `subscription: Ref<Subscription>`, обратная связь, вроде как, не нужна.

backmeupplz replied:
А если это цена, то давай назовем это `price` :)

backmeupplz replied:
И мб `price` тоже будет структурой, ведь price может быть разным, в зависимости от типа оплаты (эфир, битки, т.д.)

fanamoner committed:
add price model, change subscription model

backmeupplz commented on tsconfig.json:
```diff
@@ -11,7 +11,9 @@
     "emitDecoratorMetadata": true,
-    "experimentalDecorators": true
+    "experimentalDecorators": true,
+    "allowSyntheticDefaultImports": true,
```
А зачем нам эти изменения?

fanamoner replied:
Для исправления ошибки `TypeError: web3_1.default is not a constructor` и возможности работы с модулем Web3. Решение проблемы взял [отсюда](https://github.com/ChainSafe/web3.js/issues/2486).

backmeupplz commented on src/models/Price.ts:
```diff
@@ -0,0 +1,25 @@
+  username: string
+  @prop({ required: true })
+  chatId: number
+  @prop({ required: true, ref: 'Subscription' })
```
А можно тут сделать `ref: Subscription`? Чтобы убрать лишнюю магическую строку

backmeupplz commented on src/models/Subscription.ts:
```diff
@@ -0,0 +1,68 @@
+
+export const SubscriptionModel = getModelForClass(Subscription)
+
+export async function getOrCreateSubscription(ctx: Context) {
```
Не передавай сюда Context, пожалуйста. У этой функции не должно быть сайд-эффектов, связанных с ТГ. Она должно получать или создавать подписку. Все, что нужно сделать с ТГ — делай вне этой функции.

backmeupplz commented on src/models/Subscription.ts:
```diff
@@ -0,0 +1,68 @@
+  @prop({ required: true })
+  chatId: number
+  @prop({ required: true, unique: true })
+  addresses: { ethAddress: string }
```
Давай у `addresses` и `privateKeys` укажем тип отдельным `interface` или `class`, чтобы не городить объект тут. Можно сделать этот интерфейс или класс прямо в этом файле, экспортить его не нужно.

backmeupplz commented on src/models/Subscription.ts:
```diff
@@ -0,0 +1,68 @@
+      chatId: ctx.chat.id,
+    }).populate('subscription')
+    return ctx.reply(
+      `You already have a subscription.\nYour address is:${price.subscription.addresses.ethAddress}\nYour payment is: ${price.amount.ethAmount} ETH.`
```
Давай не различать, создана или найдена подписка. Постоянно просто говорить юзеру: для подписки на вот этот чат, вам нужно заплатить сюда вот столько.

backmeupplz commented on src/models/Subscription.ts:
```diff
@@ -0,0 +1,68 @@
+    },
+  })
+
+  const priceCreated = await PriceModel.create({
```
Слабо понимаю связь Subscription и Price. Напиши в чате, пожалуйста, что именно ты под ними понимаешь.

Я думал, что у Subscription будет просто объект `prices` типа: `{ monthly: { eth: 0.005 } }`, указывающий, сколько стоит подписка в месяц.

fanamoner committed:
Add classes for SubscriptionModel

fanamoner committed:
Remove ctx usage in getOrCreateSubscription

fanamoner committed:
Delete Price model

backmeupplz commented on src/handlers/handleStart.ts:
```diff
@@ -1,16 +1,21 @@
 
 export async function sendStart(ctx: Context) {
   const startPayload = ctx.startPayload
+  const userId = ctx.from.id
```
Если ты их только один раз испольузешь после, можно не создавать лишних переменных, а сразу исопльзовать `ctx.from.id`, например

backmeupplz commented on src/models/Subscription.ts:
```diff
@@ -0,0 +1,67 @@
+
+@modelOptions({
+  schemaOptions: { timestamps: true },
+  options: { allowMixed: Severity.ALLOW },
```
Давай выставим этот флаг сразу всей монге, а не только этой модели

backmeupplz commented on src/models/Subscription.ts:
```diff
@@ -0,0 +1,67 @@
+  const price = 1
+
+  if (subscription) {
+    return `To subscribe to this chat, you need to pay ${subscription.prices.monthly.eth} ETH to this adress: ${subscription.addresses.eth}.`
```
Выглядит так, как будто ты можешь из этой функции вернуть `subscription` и использовать его вне этой функции. Уж точно не надо возвращать строку :)

backmeupplz commented on src/models/Subscription.ts:
```diff
@@ -0,0 +1,67 @@
+export const SubscriptionModel = getModelForClass(Subscription)
+
+export async function getOrCreateSubscription(userId: number, chatId: number) {
+  const subscription = await SubscriptionModel.findOne({
```
Если сделаешь `let`, то потом сможешь не создавать `subscriptionCreated` переменную, а использовать эту

backmeupplz commented on src/models/Subscription.ts:
```diff
@@ -0,0 +1,67 @@
+    return `To subscribe to this chat, you need to pay ${subscription.prices.monthly.eth} ETH to this adress: ${subscription.addresses.eth}.`
+  }
+
+  const web3 = new Web3(Web3.givenProvider || 'ws://localhost:8545')
```
Давай вот это добавим в `.env` с записью в readme и `.env.sample`

backmeupplz replied:
Можем, кстати, web3 не создавать каждый раз, а создать один раз и экспортнуть, например, в `helpers/web3`? :)

backmeupplz commented on src/models/Subscription.ts:
```diff
@@ -0,0 +1,67 @@
+  const web3 = new Web3(Web3.givenProvider || 'ws://localhost:8545')
+  const ethAccount = web3.eth.accounts.create()
+
+  const subscriptionCreated = await SubscriptionModel.create({
```
Давай вместо этого использовать `new SubscriptionModel({ ... }).save()`

backmeupplz commented on src/models/Subscription.ts:
```diff
@@ -0,0 +1,67 @@
+    privateKeys: {
+      eth: ethAccount.privateKey,
+    },
+    prices: { monthly: { eth: price } },
```
Не нужно сразу создавать `price`, помнишь, нам нужно будет дать админу настроить цену в месяц?

fanamoner committed:
Refactor sendStart function

fanamoner committed:
Change return value of getOrCreateSubscription function

fanamoner committed:
refactor getOrCreateSubscription function

fanamoner committed:
Set global options for every mongoDB model

fanamoner committed:
Add web3 helper

fanamoner committed:
Add information about Web3 to .env.sample and README

backmeupplz commented on src/models/Subscription.ts:
```diff
@@ -0,0 +1,64 @@
+  addresses: Adresses
+  @prop({ required: true, unique: true })
+  privateKeys: PrivateKeys
+  @prop({})
```
Не нужно писать `{}`, если нет настроек, ну и он, похоже, optional, так что укажи `prices?`

backmeupplz commented on src/models/Subscription.ts:
```diff
@@ -0,0 +1,64 @@
+  subscription = await SubscriptionModel.create({
+    userId,
+    chatId,
+    addresses: {
```
Слушай, а мы можем как-то сюда прямо `ethAccount` сохранить? Или там не просто структура, а еще и методы у объекта `ethAccount`? Тип `Account` должен быть доступен в web3, вроде как

fanamoner replied:
Можно сохранить. `web3.eth.accounts.create()` возвращает объект с адресом, ключом и функциями `sign()` для транзакций и `encrypt()` . Источник: [документация Web3](https://web3js.readthedocs.io/en/v1.3.4/web3-eth-accounts.html#create)

fanamoner replied:
Функции `web3.eth.accounts.signTransaction`, `web3.eth.accounts.sign`,  `web3.eth.accounts.encrypt` существуют отдельно, поэтому их можно будет потом использовать

backmeupplz commented on src/models/Chat.ts:
```diff
@@ -10,9 +10,7 @@ export class Chat {
 
 // Get Chat model
-const ChatModel = getModelForClass(Chat, {
-  schemaOptions: { timestamps: true },
```
Зачем это убирать? Добавь таймстампы и в `Subscription`

fanamoner replied:
Я ведь выставил global options в typegoose. Все модели будут иметь `schemaOptions: { timestamps: true }`

backmeupplz replied:
Понял, спасибо!

backmeupplz commented on .env.sample:
```diff
@@ -1,2 +1,3 @@
-MONGO=mongodb://localhost:27017/tlgcoin
\ No newline at end of file
+MONGO=mongodb://localhost:27017/tlgcoin
+WEB3_PROVIDER=http://localhost:27017
```
А это валидный URL для провайдера?

fanamoner replied:
Да. Если пользователь не хочет использовать third-party провайдеров (напр. Infura), он может сам создать свой node и подключить его к локалхосту. Источник: [stackexchange](https://ethereum.stackexchange.com/questions/84582/how-to-connect-ethereum-local-provider-https-localhost8545).

backmeupplz commented on src/handlers/handleStart.ts:
```diff
@@ -10,7 +11,12 @@ export async function sendStart(ctx: Context) {
   // if no language detected, ask for language first
   if (!ctx.dbchat.language) {
-    return sendLanguage(MessageAfterLanguage.start, startPayload)(ctx)
+    sendLanguage(MessageAfterLanguage.start, startPayload)(ctx)
```
Если у человека не установлен язык, нужно этот язык спросить — а после выбора языка отправить то, что хотели отправить тут.

backmeupplz commented on src/handlers/handleStart.ts:
```diff
@@ -10,7 +11,12 @@ export async function sendStart(ctx: Context) {
   if (!ctx.dbchat.language) {
-    return sendLanguage(MessageAfterLanguage.start, startPayload)(ctx)
+    sendLanguage(MessageAfterLanguage.start, startPayload)(ctx)
+    const subscriptionInfo = await getOrCreateSubscription(
```
Это же не `subscriptionInfo`, а полноценный `subscription`

backmeupplz commented on src/handlers/handleStart.ts:
```diff
@@ -10,7 +11,12 @@ export async function sendStart(ctx: Context) {
+      ctx.from.id,
+      ctx.chat.id
+    )
+    return ctx.reply(subscriptionInfo)
```
Отправлять пользователю голый json объект плохо, всегда отправляй сразу нормальные строки, прогоняя их через i18n.

fanamoner committed:
Make price property optional

fanamoner committed:
Refactor sendStart function

Add data to i18n to reply in different languages

fanamoner committed:
Refactor subscription model structure

backmeupplz commented on locales/en.yaml:
```diff
@@ -5,3 +5,4 @@ language: Please, select the language.
 language_selected: <b>Bot</b> now speaks English.
 notifications_on: Notifications enabled
 notifications_off: Notifications disabled
+subscription_message: In order to subscribe to this group, you need to transfer 1 ETH to ${subscriptionAddress}.
```
Количество ETH тоже нужно передавать сюда переменной, оно будет разное для разных подписок

backmeupplz commented on locales/ru.yaml:
```diff
@@ -44,3 +44,4 @@ start_group_admin: Администратор группы подтвержде
 start_group_not_admin: Простите, эта ссылка существует только для администраторов сообществ.
 notifications_on: Уведомления включены
 notifications_off: Уведомления отключены
+subscription_message: Чтобы подписаться на этоу группу, вам нужно перевести 1 ETH по адресу ${subscriptionAddress}.
```
Давай опечатку в слове "этоу" поправим

backmeupplz commented on src/models/Subscription.ts:
```diff
@@ -0,0 +1,56 @@
+  monthly: { eth: number }
+}
+
+setGlobalOptions({
```
Давай это уберем в `src/models/index.ts`

backmeupplz commented on src/models/Subscription.ts:
```diff
@@ -0,0 +1,56 @@
+
+export async function getOrCreateSubscription(userId: number, chatId: number) {
+  let subscription = await SubscriptionModel.findOne({
+    chatId,
```
Надо, наверное, и по userId тоже искать?

fanamoner replied:
Если пользователь, который не является админом, будет получать подписку то будет ошибка, ведь подписка создается с `userId` админа. Я подумал, что после создания можно искать просто по `chatId`.

backmeupplz commented on src/handlers/handleStart.ts:
```diff
@@ -20,11 +21,25 @@ export async function sendStart(ctx: Context) {
     const allowedStatuses = ['administrator', 'creator']
     if (allowedStatuses.includes(chatMemberInfo.status)) {
-      return ctx.reply(ctx.i18n.t('start_group_admin'))
+      ctx.reply(ctx.i18n.t('start_group_admin'))
```
Не стоит терять `await` в функциях, которые возвращают промисы.

backmeupplz commented on src/handlers/handleStart.ts:
```diff
@@ -20,11 +21,25 @@ export async function sendStart(ctx: Context) {
     if (allowedStatuses.includes(chatMemberInfo.status)) {
-      return ctx.reply(ctx.i18n.t('start_group_admin'))
+      ctx.reply(ctx.i18n.t('start_group_admin'))
+      const subscription = await getOrCreateSubscription(
```
Здесь не нужно создавать подписку, сюда можем зайти только для конфигурации группы адмнистратором группы

backmeupplz commented on src/handlers/handleStart.ts:
```diff
@@ -20,11 +21,25 @@ export async function sendStart(ctx: Context) {
   }
 
-  return ctx.replyWithHTML(ctx.i18n.t('help'))
+  const subscription = await getOrCreateSubscription(ctx.from.id, ctx.chat.id)
```
`ctx.from.id` и `ctx.chat.id` тут будут одинаковыми, ведь это в личке :) тебе нужно из `startPayload` получать id группы, на которую хочет подписаться пользователь, если хочешь показать ему тут подписку. `startPayload` — это часть ссылки после `?start=`, например: `https://t.me/paid_subscription_bot?start=12345`

fanamoner committed:
Add subscriptionPrice variable for locales

fanamoner committed:
Refactor database structure

fanamoner committed:
Delete setGlobalOptions

setGlobalOptions did not work because of the call order. Add modelOptions instead.

fanamoner committed:
Change sendStart function

Get chatId from startPayload instead of ctx.chat.id. Make chatId type union because startPayload returns a string.

Add locales data for 'start_group_admin'.

backmeupplz commented on src/models/Subscription.ts:
```diff
@@ -0,0 +1,51 @@
+  monthly: { eth: number }
+}
+
+@modelOptions(typegooseOptions)
```
А почему глобально всем моделям не сделать эти опции? Так же и было раньше

fanamoner replied:
Оно не будет работать. Нужно, чтобы `setGlobalOptoions` применялись перед созданием всех моделей, а если они (модели) в отдельном файле, то придется в каждый файл вставлять `setGlobalOptoions`, как я понял. Смотрел [здесь](https://github.com/typegoose/typegoose/issues/128)

backmeupplz replied:
Окей. Попробуй сделать это не в `models/index`, а прямо в `app.ts:7` объявить глобальные опции. Если не получится и все равно нужно будет создавать `typegooseOptions`, то вынеси их в отдельный файл в хелперах

backmeupplz commented on src/handlers/handleStart.ts:
```diff
@@ -17,14 +18,24 @@ export async function sendStart(ctx: Context) {
     const chatId = +startPayload.replace('admin', '')
     const userId = +ctx.from.id
     const chatMemberInfo = await ctx.telegram.getChatMember(chatId, userId)
-
```
кажется, лишнее изменение

backmeupplz commented on src/handlers/handleStart.ts:
```diff
@@ -17,14 +18,24 @@ export async function sendStart(ctx: Context) {
     const chatId = +startPayload.replace('admin', '')
     const userId = +ctx.from.id
```
Хм, тут же не нужен +, вроде как. Это уже number

fanamoner committed:
Refactor database structure

Add price property to chat model. Refactor getOrCreateSubscription function and sendStart function.

backmeupplz commented on src/models/Subscription.ts:
```diff
@@ -0,0 +1,51 @@
+
+export async function getOrCreateSubscription(
+  userId: number,
+  chatId: number | string
```
Тут нужно принимать только `number`

backmeupplz commented on src/models/Subscription.ts:
```diff
@@ -0,0 +1,51 @@
+    return subscription
+  }
+
+  const chat = await ChatModel.findOne({ id: chatId })
```
А если такого чата нет?

backmeupplz commented on src/models/Subscription.ts:
```diff
@@ -0,0 +1,51 @@
+  @prop({ required: true })
+  userId: number
+  @prop({ required: true })
+  chatId: number | string
```
Делаем только `number`

backmeupplz commented on src/models/Subscription.ts:
```diff
@@ -0,0 +1,51 @@
+}
+
+@modelOptions(typegooseOptions)
+export class Subscription {
```
Добавь `{ index: true }` в `userId`, `chatId`, `chat`. Посмотри, что такое индексация в монго.

backmeupplz commented on src/handlers/handleStart.ts:
```diff
@@ -15,16 +16,27 @@ export async function sendStart(ctx: Context) {
   }
 
-  return ctx.replyWithHTML(ctx.i18n.t('help'))
+  const subscription = await getOrCreateSubscription(ctx.from.id, startPayload)
```
А что, если `startPayload` — `''`? Тогда нужно прислать просто `help`.

backmeupplz commented on src/handlers/handleStart.ts:
```diff
@@ -15,16 +16,27 @@ export async function sendStart(ctx: Context) {
+  return ctx.reply(
+    ctx.i18n.t('subscription_message', {
+      subscriptionAddress: subscription.accounts.eth.address,
+      subscriptionPrice: subscription.chat.price.monthly.eth,
```
Что, если `price` все еще не указан у чата?

fanamoner committed:
Add help message in sendStart function

fanamoner committed:
Remove union type in chatId property

fanamoner committed:
Add typegoose helper

Set global options for mongo models

fanamoner committed:
Create indexes for subscription model

fanamoner committed:
Add price check for subscription message

fanamoner committed:
Refactor getOrCreateSubscription function

backmeupplz commented on src/helpers/typegoose.ts:
```diff
@@ -0,0 +1,6 @@
+export const typegooseOptions = {
+  schemaOptions: { timestamps: true },
+  options: { allowMixed: Severity.ALLOW },
+}
```
А можем вычленить тип конфига из `typegoose`, чтобы получить тут автокомплит? И сделать `as TypegooseOptions` или что-то типа того

fanamoner replied:
Попробовал. Взял интерфейсы из [репозитория typegoose](https://github.com/typegoose/typegoose/blob/master/src/types.ts#L385).

backmeupplz commented on src/models/Chat.ts:
```diff
@@ -1,18 +1,24 @@
-import { prop, getModelForClass } from '@typegoose/typegoose'
+import { prop, getModelForClass, modelOptions } from '@typegoose/typegoose'
+import { typegooseOptions } from '@/helpers/typegoose'
```
Давай-ка назовем файл `typegooseOptions`

backmeupplz commented on src/handlers/handleStart.ts:
```diff
@@ -13,18 +14,41 @@ export async function sendStart(ctx: Context) {
     return sendLanguage(MessageAfterLanguage.start, startPayload)(ctx)
   }
 
+  if (startPayload === '') {
```
Это можно заменить на `!startPayload`

backmeupplz replied:
Если ты тут проверяешь на отсутствие пейлоада, то после этого не нужно проверять его наличие

backmeupplz commented on src/handlers/handleStart.ts:
```diff
@@ -13,18 +14,41 @@ export async function sendStart(ctx: Context) {
-  return ctx.replyWithHTML(ctx.i18n.t('help'))
+  const subscription = await getOrCreateSubscription(
+    ctx.from.id,
+    Number(startPayload)
```
Это можно заменить на `+startPayload`

backmeupplz commented on src/handlers/handleStart.ts:
```diff
@@ -13,18 +14,41 @@ export async function sendStart(ctx: Context) {
   }
 
-  return ctx.replyWithHTML(ctx.i18n.t('help'))
+  const subscription = await getOrCreateSubscription(
```
Что, если у нас появятся новые типы `startPayload` позже? Сделай проверку на то, какой это пейлоад. Вроде как, должны быть только отрицательные цифры, используй regex

fanamoner committed:
Refactor sendStart function and typegoose options

backmeupplz commented on src/handlers/handleStart.ts:
```diff
@@ -13,18 +14,44 @@ export async function sendStart(ctx: Context) {
     }
   }
 
+  if (startPayload.match(/^-?\d+$/)) {
```
А подписываться же мы только на группы и каналы можем? Скорее всего, всегда будет отрицательный тогда id

backmeupplz commented on src/handlers/handleStart.ts:
```diff
@@ -13,18 +14,44 @@ export async function sendStart(ctx: Context) {
+      +startPayload
+    )
+
+    if (subscription.chat.price) {
```
Давай сначала проверим `!subscription.chat.price`, а потом закинем вариант без ошибок. Сначала нужно ошибки проверять (типа отсутствия цены), потом в конце подходить к happy path

backmeupplz commented on src/helpers/typegooseOptions.ts:
```diff
@@ -0,0 +1,20 @@
+import { Severity } from '@typegoose/typegoose'
+import mongoose from 'mongoose'
+
+interface TypegooseOptions {
```
Нет смысла самим прописывать объект, который съедает `@modelOptions`. Используй `IModelOptions` от typegoose

fanamoner committed:
Refactor typegooseOptions.ts

fanamoner committed:
Refacor sendStart function
